fendermint/vm/topdown/src/finality.rs

```diff
@@ -0,0 +1,463 @@
+        Ok(h)
+    }
+
+    fn latest_nonce(&self) -> StmDynResult<Option<Nonce>> {
+        let top_down_msgs = self.parent_view_data.top_down_msgs.read()?;
+        Ok(top_down_msgs.upper_bound())
+    }
+
+    fn new_block_height(
+        &self,
+        height: BlockHeight,
+        block_hash: BlockHash,
+        validator_set: ValidatorSet,
+    ) -> StmDynResult<()> {
+        let insert_res = self.parent_view_data.height_data.modify(|mut cache| {
+            let r = cache.append(height, (block_hash.clone(), validator_set.clone()));
+            (cache, r)
+        })?;
+
+        match insert_res {
+            Ok(_) => Ok(()),
+            Err(_) => {
+                // now the inserted height is not the next expected block height, could be a chain
+                // reorg if the caller is behaving correctly.
+                abort(Error::ParentReorgDetected(height))
+            }
+        }
+    }
+
+    fn new_top_down_msgs(&self, top_down_msgs: Vec<CrossMsg>) -> StmDynResult<()> {
+        if top_down_msgs.is_empty() {
+            // not processing if there are no top down msgs
+            return Ok(());
+        }
+
+        ensure_sequential_by_nonce(&top_down_msgs)?;
+
+        // get the min nonce from the list of top down msgs and purge all the msgs with nonce
+        // about the min nonce in cache, as the data should be newer and more accurate.
+        let min_nonce = top_down_msgs.first().unwrap().msg.nonce;
+        self.parent_view_data.top_down_msgs.update(|mut cache| {
+            cache.remove_key_above(min_nonce);
+
+            for msg in top_down_msgs.clone() {
+                // safe to unwrap, as the append is sequential
+                cache.append(msg.msg.nonce, msg).unwrap();
+            }
+            cache
+        })?;
+
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl ParentFinalityProvider for DefaultFinalityProvider {
+    fn last_committed_finality(&self) -> StmDynResult<IPCParentFinality> {
+        let finality = self.last_committed_finality.read_clone()?;
+        Ok(finality)
+    }
+
+    fn next_proposal(&self) -> StmDynResult<IPCParentFinality> {
+        let latest_height = if let Some(h) = self.parent_view_data.latest_height()? {
+            h
+        } else {
+            return abort(Error::HeightNotReady);
+        };
+
+        // latest height has not reached, we should wait or abort
+        if latest_height < self.config.chain_head_delay {
+            return abort(Error::HeightThresholdNotReached);
+        }
+
+        let height = latest_height - self.config.chain_head_delay;
+
+        let height_data = self.parent_view_data.height_data.read()?;
+        let (block_hash, validator_set) = if let Some(v) = height_data.get_value(height) {
+            v.clone()
+        } else {
+            return abort(Error::HeightNotFoundInCache(height));
+        };
+
+        let top_down_msgs = self.parent_view_data.all_top_down_msgs()?;
```

Comment:
Unless I am missing something, I don't think we should propose all top-down messages, but the diff between the previous proposal and the new one. 
- We should track the latest nonce, `n`, seen in the proposal for the latest block committed.
- This proposal should only include the messages from `n` to the latest top-down message at the height being proposed.

As a reminder: 
- The cache should keep all the top-down messages that are final and haven't been executed yet, which means that we should check the `applied_top_down_message`  to determine what can be garbage collected.

I don't  think we should be proposing all the `top_down_messages` all the time.

Reply:
`all_top_down_msgs` returns all the top down messages stored in the cache that are not proposed/final yet.